File: 2022/day-12/day12.py

from typing import *
from collections import deque

Grid = List[List[int]]
Point = Tuple[int, int]


def parse_input(inp: str) -> Tuple[Grid, Point, Point]:
    grid = []
    start, end = None, None
    for i, line in enumerate(inp.splitlines()):
        grid.append([])
        for j, char in enumerate(line):
            if char == 'S':
                start = (i, j)
                char = 'a'
            elif char == 'E':
                end = (i, j)
                char = 'z'
            grid[i].append(ord(char) - ord('a') + 1)

    return grid, start, end


def add(a: Point, b: Point) -> Point:
    ay, ax = a
    by, bx = b
    return ay + by, ax + bx


def find_path(grid: Grid, start: Point, end: Point) -> Optional[int]:
    ny, nx = (len(grid), len(grid[0]))
    seen: Set[Point] = set()
    q: Deque[Tuple[Point, int]] = deque([(start, 0)])
    costs = dict()
    while q:
        loc, cost = q.popleft()
        if loc == end:
            return cost
        if loc in seen:
            continue
        else:
            seen |= {loc}
        for d in [(-1, 0), (0, 1), (1, 0), (0, -1)]:
            py, px = add(loc, d)
            if 0 <= py < ny and 0 <= px < nx:
                y, x = loc
                if grid[py][px] <= 1 + grid[y][x]:
                    q.append(((py, px), cost + 1))
                    costs[(px, py)] = cost + 1


def minimal_path(grid: Grid, end: Point) -> Iterable[int]:
    for i, row in enumerate(grid):
        for j, elevation in enumerate(row):
            if elevation == 1:
                d = find_path(grid, (i, j), end)
                if d is not None:
                    yield d


with open('input.txt') as h:
    content = h.read()

landscape, s, e = parse_input(content)
print('Part 1: ', find_path(landscape, s, e))
print('Part 2: ', min(minimal_path(landscape, e)))

File: 2022/day-12/day12.rs
use std::collections::{HashSet, VecDeque};

type Grid = Vec<Vec<u8>>;
type Point = (i32, i32);

fn parse_input(inp: &str) -> (Grid, Point, Point) {
    let mut start: Option<Point> = None;
    let mut end: Option<Point> = None;
    let board: Grid = inp
        .lines()
        .enumerate()
        .map(|(i, line)| {
            line.chars()
                .enumerate()
                .map(|(j, c)| {
                    let chr: char = match c {
                        'S' => {
                            start = Some((i as i32, j as i32));
                            'a'
                        }
                        'E' => {
                            end = Some((i as i32, j as i32));
                            'z'
                        }
                        'a'..='z' => c,
                        _ => panic!("Unknown elevation"),
                    };
                    ((chr as usize) - ('a' as usize)) as u8 + 1_u8
                })
                .collect()
        })
        .collect();
    (board, start.unwrap(), end.unwrap())
}

fn in_bounds(index: i32, bound: usize) -> bool {
    index >= 0 && index < bound as i32
}

fn find_path(grid: &Grid, start: Point, end: Point) -> Option<usize> {
    let (ny, nx) = (grid.len(), grid.get(0).unwrap().len());
    let mut seen: HashSet<Point> = HashSet::new();
    let mut q: VecDeque<(Point, usize)> = VecDeque::new();
    q.push_front((start, 0_usize));

    while q.len() != 0 {
        let (loc, cost) = q.pop_front().unwrap();
        if loc == end {
            return Some(cost);
        }
        if !seen.insert(loc) {
            continue;
        }
        for (dy, dx) in vec![(-1, 0), (0, 1), (1, 0), (0, -1)] {
            let (y, x) = loc;
            let (py, px) = (y + dy, x + dx);
            if in_bounds(py, ny) && in_bounds(px, nx) {
                if grid[py as usize][px as usize] <= 1 + grid[y as usize][x as usize] {
                    q.push_back(((py, px), cost + 1))
                }
            }
        }
    }
    None
}

fn minimal_path(grid: &Grid, end: Point) -> usize {
    let (ny, nx) = (grid.len(), grid.get(0).unwrap().len());
    (0..ny)
        .flat_map(move |i| (0..nx).map(move |j| (i, j)))
        .filter(|&(i, j)| grid[i][j] == 1)
        .filter_map(|(y, x)| find_path(grid, (y as i32, x as i32), end))
        .min()
        .unwrap()
}

fn main() {
    let contents = std::fs::read_to_string("input.txt").expect("Failed to read input!");
    let (grid, start, end) = parse_input(&contents);
    println!("Part 1: {:?}", find_path(&grid, start, end).unwrap());
    println!("Part 2: {:?}", minimal_path(&grid, end));
}
